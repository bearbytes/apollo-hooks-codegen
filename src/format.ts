import { PluginIR, FileIR, OperationIR, TypeIR, FragmentIR } from './types'

export function format(plugin: PluginIR): string {
  return join(
    disclaimer,
    imports,
    '',
    formatInputTypes(plugin.inputTypes),
    ...plugin.files.map(formatFile),
    '',
    boilerplate
  )
}

function formatInputTypes(inputTypes: TypeIR[]): string {
  return join(
    '',
    '',
    '/*',
    ' * GraphQL Input Types',
    ' */',
    '',
    ...inputTypes.map(formatType)
  )
}

function formatFile(file: FileIR): string {
  return join(
    '',
    '',
    '/*',
    ' * Fragments from ' + file.filePath,
    ' */',
    '',
    ...file.fragments.map(formatFragments),
    '',
    '/*',
    ' * Operations from ' + file.filePath,
    ' */',
    '',
    ...file.operations.map(formatOperation)
  )
}

function formatFragments(fragmentType: FragmentIR): string {
  return join(
    'export type ' +
      fragmentType.name +
      ' = ' +
      formatInterface(fragmentType.fields),
    '',
    ...fragmentType.fields.map(formatType),
    '',
    'const _gql_' +
      fragmentType.name +
      ' = gql`' +
      fragmentType.gqlExpression +
      '`'
  )
}

function formatOperation(operation: OperationIR): string {
  const {
    name,
    operationType,
    gqlExpression,
    data,
    variables,
    fragments,
  } = operation

  let gqlFragments = fragments
    ? join(...fragments.map(it => '${_gql_' + it + '}'))
    : ''
  let gql = 'gql`' + gqlExpression + gqlFragments + '`'

  return join(
    `export const ${name} = ${operationType}<${name}_variables, ${name}_data>(${gql})`,
    formatType(variables),
    formatType(data),
    ``
  )
}

function typeName(type: TypeIR): string {
  return [...type.namespace, type.name].join('_')
}

function formatType(type: TypeIR): string {
  const leftSide = 'export type ' + typeName(type)
  const fragments = type.fragments ? type.fragments.join(' & ') + ' & ' : ''
  const rightSide = type.scalar || formatInterface(type.fields)

  let output = leftSide + ' = ' + fragments + rightSide

  if (type.fields) {
    output += join('', ...type.fields.map(formatType))
  }

  return output
}

function formatInterface(fields: TypeIR[]) {
  return join('{', ...fields.map(field => indent(formatField(field))), '}')
}

function formatField(field: TypeIR) {
  let type = typeName(field)
  let optional = false
  if (field.modifiers) {
    optional = field.modifiers[0] == 'Nullable'
    for (const modifier of field.modifiers.reverse()) {
      type = modifier + '<' + type + '>'
    }
  }
  return field.name + (optional ? '?: ' : ': ') + type
}

const disclaimer = `
/*
 * This file was generated by graphql-code-generator with the apollo-hooks-codegen plugin.
 * Any changes made to the file will be overwritten.
 */
`

const imports = `
import * as React from 'react'
import { createContext, useEffect, useState, useContext, useRef } from 'react'
import ApolloClient, {
  MutationOptions,
  ObservableQuery,
  WatchQueryOptions,
  SubscriptionOptions,
  ApolloQueryResult,
} from 'apollo-client'
import { FetchResult, Observable } from 'apollo-link'
import { DocumentNode } from 'graphql'
import gql from 'graphql-tag'
`

const boilerplate = `
/*
 * Boilerplate
 */

type Nullable<T> = T | null
type Omit<T, K> = Pick<T, Exclude<keyof T, K>>
type QueryOpts<V> = Omit<WatchQueryOptions<V>, 'query'>
type MutateOpts<D, V> = Omit<MutationOptions<D, V>, 'mutation'>
type SubscriptionOpts<V> = Omit<SubscriptionOptions<V>, 'query'>

// We grab the ApolloClient from this context within our hooks
type ContextType = { apolloClient?: ApolloClient<any> }
const apolloContext = createContext<ContextType>({})

// Must be inserted at the root of all components that want to use the hook
// functions supplied by this file.
export function ApolloHooksProvider({
  children,
  apolloClient,
}: {
  children?: React.ReactNode
  apolloClient: ApolloClient<any> | undefined
}) {
  const elementType = apolloContext.Provider
  const elementProps: React.ProviderProps<ContextType> = {
    value: { apolloClient },
  }
  return React.createElement(elementType, elementProps, children)
}

// Executes a query that has been created by calling the exported function with
// the same name as the query operation.
// The React Hooks rules apply - this function must be called unconditionally
// within a functional React Component and will rerender the component whenever
// the query result changes.
export function useApolloQuery<D, V>(
  configuredQuery: (client: ApolloClient<any>) => ObservableQuery<D, V>,
  queryCallback?: (query: ObservableQuery<D, V>) => void
): Nullable<ApolloQueryResult<D>> {
  const { apolloClient } = useContext(apolloContext)
  if (!apolloClient) throw 'No ApolloClient provided'

  const query = useRef(configuredQuery(apolloClient))

  const [result, setResult] = useState<Nullable<ApolloQueryResult<D>>>(null)
  useEffect(() => {
    if (queryCallback) {
      queryCallback(query.current)
    }

    const subscription = query.current.subscribe(setResult)
    return () => subscription.unsubscribe()
  }, [])

  return result
}

// Prepares a mutate function when supplied with the exported function with
// the same name as the mutation operation.
// The React Hooks rules apply - this function must be called unconditionally
// within a functional React Component.
export function useApolloMutation<D, V>(
  configuredMutation: (
    client: ApolloClient<any>
  ) => (opts?: MutateOpts<D, V>) => Promise<FetchResult<D>>
) {
  const { apolloClient } = useContext(apolloContext)
  if (!apolloClient) throw 'No ApolloClient provided'
  const mutate = configuredMutation(apolloClient)
  return mutate
}

export function useApolloSubscription<D>(
  configuredSubscription: (client: ApolloClient<any>) => Observable<D>
): Nullable<D> {
  const { apolloClient } = useContext(apolloContext)
  if (!apolloClient) throw 'No ApolloClient provided'

  const observable = useRef(configuredSubscription(apolloClient))

  const [result, setResult] = useState<Nullable<D>>(null)
  useEffect(() => {
    const subscription = observable.current.subscribe(setResult)
    return () => subscription.unsubscribe()
  })

  return result
}

// Converts a gql-snippet into a user-callable function that takes options,
// which can then be passed to useApolloMutation to provide the mutate function.
function mutation<V, D>(mutation: DocumentNode) {
  return function configureMutation(opts: MutateOpts<D, V> = {}) {
    return function loadMutation(client: ApolloClient<any>) {
      return function executeMutation(opts2: MutateOpts<D, V> = {}) {
        return client.mutate<D>({ mutation, ...opts, ...opts2 })
      }
    }
  }
}

// Converts a gql-snippet into a user-callable function that takes options,
// which can then be passed to useApolloQuery to execute the query.
function query<V, D>(doc: DocumentNode) {
  return function configureQuery(opts: QueryOpts<V> = {}) {
    return function executeQuery(client: ApolloClient<any>) {
      return client.watchQuery<D>({ query: doc, ...opts })
    }
  }
}

function subscription<V, D>(doc: DocumentNode) {
  return function configureSubscription(opts: SubscriptionOpts<V> = {}) {
    return function executeSubscription(client: ApolloClient<any>) {
      return client.subscribe<D, V>({ query: doc, ...opts })
    }
  }
}
`

function join(...lines: string[]) {
  return lines.join('\n')
}

function indent(multilineText: string, offset: string = '  ') {
  return multilineText
    .trim()
    .split('\n')
    .map(line => offset + line)
    .join('\n')
}
