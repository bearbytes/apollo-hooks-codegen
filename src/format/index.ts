import { PluginIR } from '../types'
import { disclaimer, imports, boilerplate } from './Boilerplate'
import { formatInputTypes } from './InputTypes'
import { formatFile } from './File'

export function format(plugin: PluginIR): string {
  return [
    disclaimer,
    imports,
    '',
    formatInputTypes(plugin.inputTypes),
    ...plugin.files.map(formatFile),
    '',
    boilerplate,
  ].join('\n')
}
