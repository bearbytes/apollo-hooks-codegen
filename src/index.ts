import {
  PluginFunction,
  DocumentFile,
  GraphQLSchema,
  DocumentNode,
} from 'graphql-codegen-core'
import {
  DefinitionNode,
  OperationDefinitionNode,
  VariableDefinitionNode,
  SelectionSetNode,
  isListType,
  SelectionNode,
  FieldNode,
  GraphQLObjectType,
  OperationTypeNode,
  GraphQLOutputType,
  isNonNullType,
  isScalarType,
  GraphQLScalarType,
  isObjectType,
  Location,
  TypeNode,
  NamedTypeNode,
  ListTypeNode,
  GraphQLList,
  GraphQLInputObjectType,
  isInputType,
  isInputObjectType,
  GraphQLInputType,
  GraphQLInputField,
  isEnumType,
} from 'graphql'

export interface PluginConfig {
  idType?: string
  scalarTypes?: { [scalarName: string]: string }
}

export function plugin(
  schema: GraphQLSchema,
  documents: DocumentFile[],
  config: PluginConfig
) {
  const inputTypeNames: InputTypeNames = {}

  return join(
    disclaimer,
    imports,
    '',
    ...documents.map(documentFile => formatDocumentFile(documentFile)),
    '',
    formatInputNames(),
    boilerplate
  )

  function formatInputNames() {
    findRecursiveInputReferences(Object.keys(inputTypeNames))

    if (Object.keys(inputTypeNames).length == 0) return ''

    return join(
      '/*',
      ' * GraphQL InputTypes',
      ' */',
      '',
      ...Object.keys(inputTypeNames).map(name => {
        const node = schema.getType(name) as GraphQLInputObjectType
        return formatGraphQLInputObjectType(node)
      })
    )
  }

  function findRecursiveInputReferences(newNames: string[]) {
    if (newNames.length == 0) return

    const newNames2: string[] = []

    for (const name of newNames) {
      const node = schema.getType(name) as GraphQLInputObjectType
      const fieldMap = node.getFields()
      const fields = Object.values(fieldMap)
      for (const field of fields) {
        handleInputType(field.type)
      }
    }

    findRecursiveInputReferences(newNames2)

    function handleInputType(node: GraphQLInputType) {
      if (isNonNullType(node)) {
        handleInputType(node.ofType)
        return
      }
      if (isListType(node)) {
        handleInputType(node.ofType)
        return
      }
      if (isInputObjectType(node)) {
        if (!inputTypeNames[node.name]) {
          inputTypeNames[node.name] = true
          newNames2.push(node.name)
        }
        return
      }
    }
  }

  function formatGraphQLInputObjectType(node: GraphQLInputObjectType) {
    const fieldMap = node.getFields()
    const fields = Object.values(fieldMap)
    return join(
      'interface ' + node.name + '{',
      ...fields.map(field => formatInputField(field)),
      '}\n'
    )
  }

  function formatInputField(field: GraphQLInputField) {
    const isRequired = isNonNullType(field.type)
    const questionMark = isRequired ? '' : '?'
    return field.name + questionMark + ': ' + formatInputType(field.type)
  }

  function formatInputType(node: GraphQLInputType): string {
    if (isNonNullType(node)) {
      return node.ofType
    }
    if (isListType(node)) {
      return 'ReadonlyArray<' + formatInputType(node.ofType) + '>'
    }
    if (isEnumType(node)) {
      // ! TODO
      throw 'Unhandled GraphQLEnumType in formatInputType'
    }
    if (isScalarType(node)) {
      return formatGraphQLScalarType(node)
    }
    if (isInputObjectType(node)) {
      return node.name
    }
    throw 'Unhandled GraphQLInputType'
  }

  function formatDocumentFile(file: DocumentFile) {
    return join(
      '',
      '',
      '/*',
      ' * Operations from ' + file.filePath,
      ' */',
      '',
      formatDocumentNode(file.content)
    )
  }

  function formatDocumentNode(node: DocumentNode) {
    return node.definitions
      .map(definitionNode => formatDefinition(definitionNode))
      .join('\n')
  }

  function formatDefinition(node: DefinitionNode) {
    switch (node.kind) {
      case 'OperationDefinition':
        return formatOperationDefinition(node)
      default:
        throw 'unhandled DefinitionNode.kind = ' + node.kind
    }
  }

  function formatOperationDefinition(node: OperationDefinitionNode) {
    return (
      'export const ' +
      node.name.value +
      ' = ' +
      definerFunction(node.operation) +
      '<' +
      formatVariableDefinitions(node.variableDefinitions) +
      ',' +
      formatSelectionSet(
        node.selectionSet,
        selectSchemaObject(node.operation)
      ) +
      '>(gql`\n' +
      indent(formatLoc(node.loc), '  ') +
      '\n`)\n'
    )

    function definerFunction(operation: OperationTypeNode) {
      switch (operation) {
        case 'query':
          return 'defineQuery'
        case 'mutation':
          return 'defineMutation'
        case 'subscription':
          throw 'TODO'
      }
    }
  }

  function formatLoc(loc: Location) {
    return loc.source.body.substring(loc.start, loc.end)
  }

  function formatVariableDefinitions(
    nodes: ReadonlyArray<VariableDefinitionNode>
  ) {
    const list = nodes.map(variableNode =>
      formatVariableDefinition(variableNode)
    )
    return join('{', '  /* variables */', ...list, '}')
  }

  function formatVariableDefinition(node: VariableDefinitionNode) {
    // todo defaultValue

    const isRequired = node.type.kind == 'NonNullType'
    const questionMark = isRequired ? '' : '?'

    return (
      node.variable.name.value + questionMark + ': ' + formatTypeNode(node.type)
    )
  }

  function formatTypeNode(node: TypeNode): string {
    if (node.kind == 'NonNullType') {
      return formatTypeNodeNotNull(node.type)
    }
    return 'null | ' + formatTypeNodeNotNull(node)
  }

  function formatTypeNodeNotNull(node: NamedTypeNode | ListTypeNode) {
    switch (node.kind) {
      case 'ListType':
        return 'Array<' + formatTypeNode(node.type) + '>'
      case 'NamedType':
        return formatNameTypeNode(node)
    }
  }

  function formatNameTypeNode(node: NamedTypeNode) {
    switch (node.name.value) {
      case 'String':
        return 'string'
      case 'Int':
        return 'number'
      case 'Float':
        return 'number'
      case 'Boolean':
        return 'boolean'
      case 'ID':
        return config.idType || 'string'
      default:
        const nodeType = schema.getType(node.name.value).astNode.kind
        if (nodeType == 'ScalarTypeDefinition') {
          return formatCustomScalar(node.name.value)
        } else {
          inputTypeNames[node.name.value] = true
          return node.name.value
        }
    }
  }

  function formatSelectionSet(
    node: SelectionSetNode,
    schemaObject: GraphQLObjectType
  ) {
    const list = node.selections.map(selectionNode =>
      formatSelectionNode(selectionNode, schemaObject, '  ')
    )
    return join('{', '  /* data */', ...list, '}')
  }

  function formatSelectionNode(
    node: SelectionNode,
    schemaObject: GraphQLObjectType,
    offset: string
  ) {
    switch (node.kind) {
      case 'Field':
        return formatFieldNode(node, schemaObject, offset)
      default:
        throw 'unhandled SelectionNode.kind = ' + node.kind
    }
  }

  function formatFieldNode(
    node: FieldNode,
    schemaObject: GraphQLObjectType,
    offset: string
  ) {
    const schemaName = node.name.value
    const aliasName = node.alias ? node.alias.value : schemaName

    const schemaField = schemaObject.getFields()[schemaName]
    const isRequired = isNonNullType(schemaField.type)
    const questionMark = isRequired ? '' : '?'

    return (
      offset +
      aliasName +
      questionMark +
      ': ' +
      formatGraphQLOutputType(
        schemaField.type,
        node.selectionSet,
        offset + '  '
      )
    )
  }

  function formatGraphQLOutputType(
    type: GraphQLOutputType,
    selectionSet: SelectionSetNode,
    offset: string
  ): string {
    if (isNonNullType(type)) {
      return formatGraphQLOutputTypeNotNull(type.ofType, selectionSet, offset)
    }
    return (
      'null | ' + formatGraphQLOutputTypeNotNull(type, selectionSet, offset)
    )
  }

  function formatGraphQLOutputTypeNotNull(
    type: GraphQLOutputType,
    selectionSet: SelectionSetNode,
    offset: string
  ) {
    if (isScalarType(type)) {
      return formatGraphQLScalarType(type)
    }
    if (isObjectType(type)) {
      return formatGraphQLObjectType(type, selectionSet, offset)
    }
    if (isListType(type)) {
      return (
        'Array<' +
        formatGraphQLOutputType(type.ofType, selectionSet, offset) +
        '>'
      )
    }
    if (isEnumType(type)) {
      const enumValues = type.getValues()
      return enumValues.map(value => "'" + value.name + "'").join(' | ')
    }
    throw 'unhandled GraphQLOutputType "' + type + '"'
  }

  function formatGraphQLScalarType(type: GraphQLScalarType) {
    switch (type.name) {
      case 'String':
        return 'string'
      case 'Int':
        return 'number'
      case 'Float':
        return 'number'
      case 'Boolean':
        return 'boolean'
      case 'ID':
        return config.idType || 'string'
      default:
        return formatCustomScalar(type.name)
    }
  }

  function formatCustomScalar(name: string) {
    const customName = config.scalarTypes && config.scalarTypes[name]
    return customName || 'any'
  }

  function formatGraphQLObjectType(
    type: GraphQLObjectType,
    selectionSet: SelectionSetNode,
    offset: string
  ) {
    const list = selectionSet.selections.map(selectionNode =>
      formatSelectionNode(selectionNode, type, offset)
    )
    return join('{', ...list, '}')
  }

  function selectSchemaObject(operation: OperationTypeNode) {
    switch (operation) {
      case 'query':
        return schema.getQueryType()
      case 'mutation':
        return schema.getMutationType()
      case 'subscription':
        return schema.getSubscriptionType()
    }
  }
}

const disclaimer = `
/*
 * This file was generated by graphql-code-generator with the apollo-hooks-codegen plugin.
 * Any changes made to the file will be overwritten.
 */
`

const imports = `
import * as React from 'react'
import { createContext, useEffect, useState, useContext } from 'react'
import ApolloClient, {
  MutationOptions,
  ObservableQuery,
  WatchQueryOptions,
  ApolloCurrentResult,
} from 'apollo-client'
import { FetchResult } from 'apollo-link'
import { DocumentNode } from 'graphql'
import gql from 'graphql-tag'
`

const boilerplate = `
/*
 * Boilerplate
 */

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>
type Error = any
type QueryOpts<V> = Omit<WatchQueryOptions<V>, 'query'>
type MutateOpts<D, V> = Omit<MutationOptions<D, V>, 'mutation'>

// We grab the ApolloClient from this context within our hooks
type ContextType = { apolloClient?: ApolloClient<any> }
const apolloContext = createContext<ContextType>({})

// Must be inserted at the root of all components that want to use the hook
// functions supplied by this file.
export function ApolloHooksProvider({
  children,
  apolloClient,
}: {
  children?: React.ReactNode
  apolloClient: ApolloClient<any> | undefined
}) {
  const elementType = apolloContext.Provider
  const elementProps: React.ProviderProps<ContextType> = {
    value: { apolloClient },
  }
  return React.createElement(elementType, elementProps, children)
}

// Converts a gql-snippet into a user-callable function that takes options,
// which can then be passed to useApolloQuery to execute the query.
function defineQuery<V, D>(doc: DocumentNode) {
  return function configureQuery(opts: QueryOpts<V> = {}) {
    return function executeQuery(client: ApolloClient<any>) {
      return client.watchQuery<D>({ query: doc, ...opts })
    }
  }
}

// Executes a query that has been created by calling the exported function with
// the same name as the query operation.
// The React Hooks rules apply - this function must be called unconditionally
// within a functional React Component and will rerender the component whenever
// the query result changes.
export function useApolloQuery<D, V>(
  configuredQuery: (client: ApolloClient<any>) => ObservableQuery<D, V>
): [ApolloCurrentResult<D>, ObservableQuery<D, V>] {
  const { apolloClient } = useContext(apolloContext)
  if (!apolloClient) throw 'No ApolloClient provided'

  const query = configuredQuery(apolloClient)

  const [result, setResult] = useState(query.currentResult())
  useEffect(() => {
    const subscription = query.subscribe(setResult)
    return () => subscription.unsubscribe()
  }, [])

  return [result, query]
}

// Converts a gql-snippet into a user-callable function that takes options,
// which can then be passed to useApolloMutation to provide the mutate function.
function defineMutation<V, D>(mutation: DocumentNode) {
  return function configureMutation(opts: MutateOpts<D, V> = {}) {
    return function loadMutation(client: ApolloClient<any>) {
      return function executeMutation(opts2: MutateOpts<D, V> = {}) {
        return client.mutate<D>({ mutation, ...opts, ...opts2 })
      }
    }
  }
}

// Prepares a mutate function when supplied with the exported function with
// the same name as the mutation operation.
// The React Hooks rules apply - this function must be called unconditionally
// within a functional React Component.
export function useApolloMutation<D, V>(
  configuredMutation: (
    client: ApolloClient<any>
  ) => (opts?: MutateOpts<D, V>) => Promise<FetchResult<D>>
) {
  const { apolloClient } = useContext(apolloContext)
  if (!apolloClient) throw 'No ApolloClient provided'
  const mutate = configuredMutation(apolloClient)
  return mutate
}
`

function join(...lines: string[]) {
  return lines.join('\n')
}

function indent(multilineText: string, offset: string) {
  return multilineText
    .trim()
    .split('\n')
    .map(line => offset + line)
    .join('\n')
}

type InputTypeNames = { [name: string]: boolean }
