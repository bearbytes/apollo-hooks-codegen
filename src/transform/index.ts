import { GraphQLSchema } from 'graphql'
import { DocumentFile } from 'graphql-codegen-core'
import { PluginConfig, PluginIR } from '../types'
import { setGlobals, InputTypes } from './Globals'
import { transformDocumentFile } from './DocumentFile'

export function transform(
  schema: GraphQLSchema,
  documents: DocumentFile[],
  config: PluginConfig
): PluginIR {
  setGlobals(schema, config)

  const files = documents.map(transformDocumentFile)
  const inputTypes = InputTypes()

  return { files, inputTypes }
}
