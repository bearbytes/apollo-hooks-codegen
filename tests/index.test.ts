import { executeCodegen } from 'graphql-code-generator'
import { plugin } from '../src/index'
import * as fs from 'fs-extra'

test('basic test', async () => {
  const schema = await fs.readFile(__dirname + '/schema.graphql', {
    encoding: 'utf8',
  })
  const documents = await fs.readFile(__dirname + '/queries.graphql', {
    encoding: 'utf8',
  })

  const output = await executeCodegen({
    schema,
    documents,
    generates: {
      'output.ts': ['apollo-hooks-codegen'],
    },
    pluginLoader,
  })

  const fileContent = output.find(it => it.filename == 'output.ts').content
  expect(fileContent).toMatchSnapshot()
})

function pluginLoader(pluginName: string) {
  // graphql-codegen-codegen prefixes "graphql-codegen-" to the plugin
  if (pluginName == 'graphql-codegen-apollo-hooks-codegen')
    return {
      plugin,
    }
  throw 'Plugin not found'
}
