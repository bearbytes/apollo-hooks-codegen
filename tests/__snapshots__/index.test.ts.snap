// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic test 1`] = `
"/*
 * This file was generated by graphql-code-generator with the apollo-hooks-codegen plugin.
 * Any changes made to the file will be overwritten.
 */

import * as React from 'react'
import { createContext, useEffect, useState, useContext } from 'react'
import ApolloClient, {
  MutationOptions,
  ObservableQuery,
  WatchQueryOptions,
} from 'apollo-client'
import { FetchResult } from 'apollo-link'
import { DocumentNode } from 'graphql'
import gql from 'graphql-tag'

/*
 * Operations from document.graphql
 */

export const testScalars = defineQuery<
  {
    /* variables */
  },
  {
    /* data */
    scalarString?: null | string
    scalarInt?: null | number
    scalarFloat?: null | number
    scalarBoolean?: null | boolean
    scalarID?: null | string
    scalarCustom?: null | any
    renamedString?: null | string
  }
>(gql\`
  query testScalars {
    scalarString
    scalarInt
    scalarFloat
    scalarBoolean
    scalarID
    scalarCustom

    renamedString: scalarString
  }
\`)

export const testNullability = defineQuery<
  {
    /* variables */
  },
  {
    /* data */
    nullableBoolean?: null | boolean
    nonNullableBoolean: boolean
  }
>(gql\`
  query testNullability {
    nullableBoolean
    nonNullableBoolean
  }
\`)

export const testNesting = defineQuery<
  {
    /* variables */
  },
  {
    /* data */
    nestedObject?: null | {
      recursive?: null | {
        recursive?: null | {
          scalar?: null | boolean
        }
      }
      nested?: null | {
        scalar?: null | number
      }
      scalar?: null | boolean
      list?: null | Array<string>
    }
  }
>(gql\`
  query testNesting {
    nestedObject {
      recursive {
        recursive {
          scalar
        }
      }
      nested {
        scalar
      }
      scalar
      list
    }
  }
\`)

export const testMethods = defineQuery<
  {
    /* variables */
    reqParam: string
    optParam: number
    input?: null | InputType
    list2: Array<any>
  },
  {
    /* data */
    method?: null | any
    renamedMethod?: null | Array<null | any>
  }
>(gql\`
  query testMethods(
    $reqParam: String!
    $optParam: Float!
    $input: InputType
    $list2: [CustomScalar!]!
  ) {
    method(requiredParam: $reqParam, optionalParam: $optParam, input: $input)
    renamedMethod: method2(list2: $list2)
  }
\`)

export const testScalarsMutation = defineMutation<
  {
    /* variables */
  },
  {
    /* data */
    scalarString?: null | string
    scalarInt?: null | number
    scalarFloat?: null | number
    scalarBoolean?: null | boolean
    scalarID?: null | string
    scalarCustom?: null | any
    renamedString?: null | string
  }
>(gql\`
  mutation testScalarsMutation {
    scalarString
    scalarInt
    scalarFloat
    scalarBoolean
    scalarID
    scalarCustom

    renamedString: scalarString
  }
\`)

export const testNullabilityMutation = defineMutation<
  {
    /* variables */
  },
  {
    /* data */
    nullableBoolean?: null | boolean
    nonNullableBoolean: boolean
  }
>(gql\`
  mutation testNullabilityMutation {
    nullableBoolean
    nonNullableBoolean
  }
\`)

export const testNestingMutation = defineMutation<
  {
    /* variables */
  },
  {
    /* data */
    nestedObject?: null | {
      recursive?: null | {
        recursive?: null | {
          scalar?: null | boolean
        }
      }
      nested?: null | {
        scalar?: null | number
      }
      scalar?: null | boolean
      list?: null | Array<string>
    }
  }
>(gql\`
  mutation testNestingMutation {
    nestedObject {
      recursive {
        recursive {
          scalar
        }
      }
      nested {
        scalar
      }
      scalar
      list
    }
  }
\`)

export const testMethodsMutation = defineMutation<
  {
    /* variables */
    reqParam: string
    optParam: number
    input?: null | InputType
    list2: Array<any>
  },
  {
    /* data */
    method?: null | any
    renamedMethod?: null | Array<null | any>
  }
>(gql\`
  mutation testMethodsMutation(
    $reqParam: String!
    $optParam: Float!
    $input: InputType
    $list2: [CustomScalar!]!
  ) {
    method(requiredParam: $reqParam, optionalParam: $optParam, input: $input)
    renamedMethod: method2(list2: $list2)
  }
\`)

/*
 * GraphQL InputTypes
 */

interface InputType {
  recursive?: InputType
  nested?: NestedInputType
  scalar?: boolean
  list?: ReadonlyArray<String>
}

interface NestedInputType {
  scalar?: number
}

/*
 * Boilerplate
 */

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>
type Error = any
type QueryOpts<V> = Omit<WatchQueryOptions<V>, 'query'>
type MutateOpts<D, V> = Omit<MutationOptions<D, V>, 'mutation'>

// We grab the ApolloClient from this context within our hooks
type ContextType = { apolloClient?: ApolloClient<any> }
const apolloContext = createContext<ContextType>({})

// Must be inserted at the root of all components that want to use the hook
// functions supplied by this file.
export function ApolloHooksProvider({
  children,
  apolloClient,
}: {
  children: React.ReactNode
  apolloClient?: ApolloClient<any> | undefined
}) {
  const elementType = apolloContext.Provider
  const elementProps: React.ProviderProps<ContextType> = {
    value: { apolloClient },
  }
  return React.createElement(elementType, elementProps, children)
}

// Converts a gql-snippet into a user-callable function that takes options,
// which can then be passed to useApolloQuery to execute the query.
function defineQuery<V, D>(doc: DocumentNode) {
  return function configureQuery(opts: QueryOpts<V> = {}) {
    return function executeQuery(client: ApolloClient<any>) {
      return client.watchQuery<D>({ query: doc, ...opts })
    }
  }
}

// Executes a query that has been created by calling the exported function with
// the same name as the query operation.
// The React Hooks rules apply - this function must be called unconditionally
// within a functional React Component and will rerender the component whenever
// the query result changes.
export function useApolloQuery<D, V>(
  configuredQuery: (client: ApolloClient<any>) => ObservableQuery<D, V>
): [D | undefined, Error | undefined] {
  const { apolloClient } = useContext(apolloContext)
  if (!apolloClient) throw 'No ApolloClient provided'

  const watchQuery = configuredQuery(apolloClient)

  const [data, setData] = useState<D | undefined>(undefined)
  const [error, setError] = useState<Error | undefined>(undefined)
  useEffect(() => {
    const subscription = watchQuery.subscribe(
      result => setData(result.data),
      error => setError(error)
    )
    return () => subscription.unsubscribe()
  }, [])

  return [data, error]
}

// Converts a gql-snippet into a user-callable function that takes options,
// which can then be passed to useApolloMutation to provide the mutate function.
function defineMutation<V, D>(mutation: DocumentNode) {
  return function configureMutation(opts: MutateOpts<D, V> = {}) {
    return function loadMutation(client: ApolloClient<any>) {
      return function executeMutation(opts2: MutateOpts<D, V> = {}) {
        return client.mutate<D>({ mutation, ...opts, ...opts2 })
      }
    }
  }
}

// Prepares a mutate function when supplied with the exported function with
// the same name as the mutation operation.
// The React Hooks rules apply - this function must be called unconditionally
// within a functional React Component.
export function useApolloMutation<D, V>(
  configuredMutation: (
    client: ApolloClient<any>
  ) => (opts?: MutateOpts<D, V>) => Promise<FetchResult<D>>
) {
  const { apolloClient } = useContext(apolloContext)
  if (!apolloClient) throw 'No ApolloClient provided'
  const mutate = configuredMutation(apolloClient)
  return mutate
}
"
`;
